onnxruntime/core/providers/xnnpack/nn/average_pool.cc

```diff
@@ -0,0 +1,271 @@
+
+  uint32_t pooling_height = gsl::narrow<uint32_t>(pool_attrs.kernel_shape[0]);
+  uint32_t pooling_width = gsl::narrow<uint32_t>(pool_attrs.kernel_shape[1]);
+  uint32_t stride_height = gsl::narrow<uint32_t>(pool_attrs.strides[0]);
+  uint32_t stride_width = gsl::narrow<uint32_t>(pool_attrs.strides[1]);
+
+  uint32_t flags = 0;
+  if (pool_attrs.auto_pad == AutoPadType::SAME_UPPER) {
+    flags |= XNN_FLAG_TENSORFLOW_SAME_PADDING;
+  }
+
+  xnn_status status = xnn_status_unsupported_parameter;
+  if (avgpool_type == OpComputeType::op_compute_type_fp32) {
+    float output_min = clip_min_max ? clip_min_max->first : -INFINITY;
+    float output_max = clip_min_max ? clip_min_max->second : INFINITY;
+
+    status = xnn_create_average_pooling2d_nhwc_f32(input_padding_top, input_padding_right,
+                                                   input_padding_bottom, input_padding_left,
+                                                   pooling_height, pooling_width,
+                                                   stride_height, stride_width,
+                                                   C, C, C,  // channels, input_pixel_stride, output_pixel_stride
+                                                   output_min, output_max, flags, &p);
+  } else if (avgpool_type == OpComputeType::op_compute_type_qu8) {
+    uint8_t output_min = 0;
+    uint8_t output_max = 255;
+    status = xnn_create_average_pooling2d_nhwc_qu8(input_padding_top, input_padding_right,
+                                                   input_padding_bottom, input_padding_left,
+                                                   pooling_height, pooling_width,
+                                                   stride_height, stride_width,
+                                                   C, C, C,  // channels, input_pixel_stride, output_pixel_stride
+                                                   quant_param[0].second,
+                                                   quant_param[0].first[0],
+                                                   quant_param[1].second,
+                                                   quant_param[1].first[0],
+                                                   output_min, output_max, flags, &p);
+  }
+
+  if (status != xnn_status_success) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "xnn_create_average_pooling2d_nhwc_",
+                           OpTypeToString(avgpool_type), " failed. Status:", status);
+  }
+  return Status::OK();
+}
+
+static bool IsQuantAvgPoolSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+  bool supported = false;
+  do {
+    TensorQuantType x_input_type, output_type;
+
+    x_input_type = GetTensorQuantType(node_unit, 0, false, graph);
+    output_type = GetTensorQuantType(node_unit, 0, true, graph);
+    if (x_input_type != TensorTypeUint8 ||
+        output_type != TensorTypeUint8) {
+      break;
+    }
+    supported = true;
+  } while (false);
```

Comment:
nit: using `do {} while(false);` is handy when there are multiple exit points. In this case it's probably overcomplicating the code given it could simply be

```
    var x_input_type = GetTensorQuantType(node_unit, 0, false, graph);
    var output_type = GetTensorQuantType(node_unit, 0, true, graph);
    return (x_input_type == TensorTypeUint8 && output_type == TensorTypeUint8);

```